File: analysis/perf-env/printenv.c
/// Program that prints its own environment, illustrates how running perf-stat
/// adds some environment variables.
///
/// $ perf stat env -i ./printenv
///     <empty>
///
/// $ env -i perf stat ./printenv
///     PWD=/home/lars/bias-paper/code/perf-env
///     SHLVL=0
///     PERF_BUILDID_DIR=.debug
///     PATH=/usr/libexec/perf-core:/usr/lib/linux-tools/3.11.0-15-generic:/usr/local/bin:/usr/bin:/bin
///
/// $ perf stat ./printenv
///     PWD=/home/lars/bias-paper/code/perf-env
///     SHLVL=0
///     PERF_BUILDID_DIR=.debug
///     PATH=/usr/libexec/perf-core:/usr/lib/linux-tools/3.11.0-15-generic:/usr/local/bin:/usr/bin:/bin
///     (+ whatever was already there)
///
#include <stdio.h>
#include <string.h>

int main(int argc, char *argv[]) {
    long i = 0, *addr;
    char file[64], command[128];

    for (; i < argc; ++i) 
        puts(argv[i]);
    puts("");

    sprintf(file, "/proc/%d/environ", getpid());

    strcpy(command, "cat ");
    strcat(command, file);
    system(command);

    puts("");

    strcat(command, " | wc -c");
    system(command);

    asm("movq %%rbp, %0" : "=r"(addr));
    printf("%%rbp: %p\n", (void*)&addr);

    return 0;
}

File: analysis/perf-env/printenv.py
#!/usr/bin/env python 

import subprocess

# Generating plots with lplot are done using python and subprocess. This script
# is used to invoke printenv, to show how subprocess.open with an empty
# environment can possibly affect addresses compared to executing from shell.
# In conclusion, perf stat adds about 170 bytes, including PWD which will
# obviously vary between setups. Launching through subprocess.Popen does not
# affect environment or variable addresses. 

# $ ./printenv.py
# PWD=/home/lars/bias-paper/code/perf-env
# SHLVL=0
# PERF_BUILDID_DIR=.debug
# PATH=/usr/libexec/perf-core:/usr/lib/linux-tools/3.11.0-15-generic:/usr/local/bin:/usr/bin:/bin

p = subprocess.Popen("perf stat -e cycles:u bin/printenv", env={}, shell=True)
p.wait()
